FEP0002-Asynchronous-Document-Recompute-and-Multithreading-Infrastructure.md

```diff
@@ -0,0 +1,144 @@
+| **Created**        | 2025-05-20                                                        |
+| **Updated**        | 2025-05-20                                                        |
+| **Discussion**     | TBD                                                               |
+| **Implementation** | TBD                                                               |
+
+A proposal to introduce foundational asynchronous and multithreaded infrastructure in FreeCAD, improving UI responsiveness and paving the way for parallel document recomputation.
+
+## Context
+
+FreeCAD’s code runs mostly entirely on the main thread: Python scripts, Qt event loop (including rendering and event dispatch), and OCCT operations all execute in sequence. When a long-running OCCT computation occurs during Qt’s event loop, it blocks incoming events—pausing UI updates, preventing user interaction, and stalling rendering until the operation completes. Currently, the only way to stop such a blocking computation is to kill the application outright, resulting in potential data loss and non-optimal user experience.
+
+## Motivation
+
+FreeCAD’s current document recompute and signal system operate synchronously, often blocking the UI during heavy OCCT computations.
+
+As models grow in complexity, users experience lags and freezes that impede productivity. By offloading compute-intensive tasks into asynchronous and multithreaded workflows, we can maintain a responsive interface and prepare the codebase for future parallelization of the document DAG.
+
+## Rationale
+
+The core driver of this proposal, has been to run feature transformation commands (e.g., Part and PartDesign operations) in the background, ensuring the UI remains interactive, while making minimal changes to the data model, as well as allowing parallel creation of FreeCAD objects, as long as they don't touch shared state.
+
+## Specification
+
+### Document Recompute Thread
+
+Introduce a dedicated recompute worker thread that will accept and process document recompute requests in FIFO order. Most common recompute requests originate from:
+
+* **Refresh commands**: User-initiated UI actions that trigger a full document recompute.
+* **Python document recompute requests**: Scripts invoking `App::Document::recompute()`.
+* **Feature transformation commands**: Part and PartDesign operations that modify features and require recompute.
+* **Property changes**: When data changes, `onChange` signal triggers recomputation.
+
+The recompute thread will handle:
+
+1. Receiving and queuing requests.
+2. Executing recompute logic off the main thread.
+3. Handling abort signals to cancel in-progress operations.
+4. Posting completion events back to the Qt main loop for UI update.
+
+### Concurrency Constraints
+
+While recompute is running, other code paths in the main thread may post events or callbacks that invoke FreeCAD APIs, including :
```

Comment:
Please add that user UI actions can trigger FreeCAD API use:
1. Qt UI user interactions
2. Coin callbacks for user actions in the 3D viewport

Diong this while a recompute worker is running may crash the application.

Consequence: FreeCAD must ensure that UI is blocked and no such events can occure due to user interaction
Alternative: Code called from those interactions must not use any FreeCAD API or data. This includes addons, which can add custom callbacks to both systems.